### ISSUE TITLE
Irritating default behavior of Operations class - raising FuseOSError(EROFS) where it really should not

### ISSUE THREAD

**Issue by s-m-e** [fusepy/fusepy#81]
This issue can be understood as a suggestions with respect to #80. 

---

The "usual procedure" for writing a filesystem against the high level API begins with sub-classing from ``Operations``. While ``libfuse`` follows the philosophy of "do not implement what you do not need", the original ``Operations`` class in ``fusepy`` contains an implementation for virtually every method supported by FUSE's high level API and adds a non-standard behavior to them. It appears that many methods return ``EROFS`` in one way or the other, indicating a read-only filesystem. 

This approach becomes highly problematic when a developer follows the guidelines of ``libfuse`` by leaving unwanted methods unimplemented. For instance, a read-write filesystem can work perfectly fine without the ``create`` method if the filesystem already offers properly implemented ``mknod`` and ``open`` methods. Nevertheless, FUSE will attempt to call ``create`` first, if it is implemented. If it is not implemented, FUSE will subsequently instead call the equivalent sequence of ``getattr``, ``mknod`` and ``open``. ``fusepy`` breaks this behavior by offering a ``create`` method by default, which will raise ``FuseOSError(EROFS)`` (read-only file system). As per the guidelines, it should either not exist or at least raise ``FuseOSError(errno.ENOSYS)`` (which does in fact work just fine). In a read-write filesystem this causes quite literally havoc and is, if ones does not know what to look for, fairly hard to debug.

Looking at the code of the ``Operations`` class, my impression is that it is designed to make it as easy as possible to write a simple read-only filesystem. It's apparently its primary focus, completely ignoring all other use-cases of FUSE. I felt into this trap and searched errors in my code for ever before I figured out that ``fusepy`` was really doing something it should not do - falsely assuming it would follow the design patterns of the library it is offering bindings for.

As a developer, the current way out of this mess is by not sub-classing from ``Operations``. When doing so, one must implement the ``__call__`` method which is required for ``fusepy`` to work correctly (copy&paste does the job ...). Alternatively, one can subclass from ``Operations`` but, instead, has to overwrite (!) every single unwanted method and raise ``FuseOSError(errno.ENOSYS)`` when they are called. Either way, the situation is not satisfying. 

Solving this without breaking backwards compatibility ... There should be a "clean" ``Operations`` class to subclass from (probably under another/new name), containing nothing but the ``__call__`` method. Everything else can be "moved" into something like a ``Operations_template_RO_mixin`` class along the lines of the ``LoggingMixIn`` class. 

When offering a clean ``Operations`` class, it is probably also a good idea to "zero-out" the ``fuse_operations`` ctypes structure ``fuse_ops`` after its initialization and before it is filled with pointers to methods.

**Comment by eestrada** [comment]
> `fusepy` breaks this
> behavior by offering a create method by default, which will raise
> FuseOSError(EROFS) (read-only file system). As per the guidelines, it should
> either not exist or at least raise FuseOSError(errno.ENOSYS) (which does in
> fact work just fine).

Based on what you have said, I think it is best to implement a class
like `OperationsBase`
which another "new" class `ReadOnlyOperations` would subsequently
inherit from and override any
methods with read-only equivalents (this would just be the old
`Operations` class with a new name).
The new `Operations` class would simply be an alias to
`ReadOnlyOperations`. This way, logging code
and tracebacks would spew information about `ReadOnlyOperations`. The
`Operations` class
would also be deprecated and the Minor version number incremented to
indicate the deprecation.

With what you said about `FuseOSError(errno.ENOSYS)` working, I would
simply replace any and
all instances of `FuseOSError(EROFS)` with `FuseOSError(errno.ENOSYS)`
in the `OperationsBase`
class. Why? Because the base class code helps indicate to new users
what methods they need
to implement to get a working FS. It is a form of passive
documentation. I know I looked at the
`Operations` class implementation in the code when I first tried to
learn what to do with the library.

**Comment by s-m-e** [comment]
> I know I looked at the `Operations` class implementation in the code when I first tried to learn what to do with the library.

I ended up looking into it because neither ``libfuse`` nor ``fusepy`` were offering any documentation worth mentioning - which is in itself virtually unacceptable, but that's a different story. 

> Because the base class code helps indicate to new users what methods they need to implement to get a working FS. 

I'd agree ... to a degree ;) For a lot of use cases, only a tiny subset of the available API is required for implementing a working filesystem. This is what the examples teach us (standing in for the missing documentation). Offering an implementation for every available API call, doing nothing but raising ``FuseOSError(errno.ENOSYS)`` (if I understand you correctly?), does not help very much from my point of view. 

> The new `Operations` class would simply be an alias to `ReadOnlyOperations`. This way, logging code and tracebacks would spew information about `ReadOnlyOperations`.

Giving hints like this through tracebacks is a really good idea.

> The `Operations` class would also be deprecated and the Minor version number incremented to indicate the deprecation.

This is one more thing on the "deprecation" list ... I feel uneasy about the entire concept of deprecation warnings and any sort of version number bump related to it. The problem is ... a lot of old code depends on ``fusepy`` as it is, with all its quirks. Most of it is unlikely to be ever fixed. So you deprecate a feature / behavior, bump the version, release - things are fine. Then you remove the feature or change the original default behavior, bump the version again, release again - people pull the package as a dependency from PiPY and whatever relies on ``fusepy`` breaks. This just is not good. It's the reason why I prefer the idea of breaking the development into two branches - one for maintaining and fixing ``fusepy`` while maintaining all of its original default "quirkyness" and one for rebuilding it without quirks in a clean, maintainable and extendable way, under a new name ... ``fusepy3`` or ``fusepy_ng`` or whatever (I am really bad at naming things) ... while the old branch can rest in peace under its old name on PyPI. The old branch can have a ``__features__`` dict (as discussed in the nanosecond opt-in scenario), which can be used for controlling some of the quirks, while the new branch could offer a significantly cleaner design - without a ``__features__`` dict, of cause. How about that?

**Comment by eestrada** [comment]
> This is one more thing on the "deprecation" list ... I feel uneasy about the entire concept of deprecation warnings and any sort of version number bump related to it. The problem is ... a lot of old code depends on fusepy as it is, with all its quirks. Most of it is unlikely to be ever fixed. So you deprecate a feature / behavior, bump the version, release - things are fine. Then you remove the feature or change the original default behavior, bump the version again, release again - people pull the package as a dependency from PiPY and whatever relies on fusepy breaks. This just is not good. It's the reason why I prefer the idea of breaking the development into two branches - one for maintaining and fixing fusepy while maintaining all of its original default "quirkyness" and one for rebuilding it without quirks in a clean, maintainable and extendable way, under a new name ... fusepy3 or fusepy_ng or whatever (I am really bad at naming things) ... while the old branch can rest in peace under its old name on PyPI. The old branch can have a __features__ dict (as discussed in the nanosecond opt-in scenario), which can be used for controlling some of the quirks, while the new branch could offer a significantly cleaner design - without a __features__ dict, of cause. How about that?

Thanks for the food for thought. There is a lot to address here and it
is late where I am. I'll do my best to remain coherent as I type. :)

TL;DR There isn't an easy answer(i.e. "it depends"). We should get
testing in place first regardless of what we choose moving forward.

I agree that version numbers are an imperfect way to address library
changes. Sadly, it is the only real mechanism we have for determining
which dependencies to pull in via `pip`. Also, it is simply the status
quo, for better or for worse; it is what people expect.

My current plan is to never bump the major version number (i.e. make a
breaking change) if I can help it. If you feel such drastic changes
are necessary that it requires breaking backward compatibility, then I
would suggest a full-on hard fork into a separate repo altogether, ala
`fusepy_ng` or `fusepyllow`. However, I think that will only cause
more confusion. "Do I use fusepy or fusepy3/fusepy_ng? They both use
ctypes. fusepy has more users, but fusepy3 says it is newer." It will
either slowly kill fusepy2 or fusepy3 will be stillborn. I think at
this point though, fusepy2.x has hit critical mass; it won't ever
really die. There are many projects now using it that are more popular
than fusepy itself. Some of them are actively maintained, some are
not. In either case, they are still heavily used. By forking,
downstream projects will assume that fusepy2 is dead and will simply
add it as a static dependency directly in their code base (as a few
have already done), thus further fracturing the fusepy landscape. Or,
the unmaintained ones will stay locked off on a version they used to
begin with.

From what you have said, if we raise FuseOSError(errno.ENOSYS), we
will just attempt multiple fallbacks until we ultimately fail, which
will have the same effect as we have now of having a read-only FS. In
that case, we can just change the behavior to use errno.ENOSYS, not
worry about deprecating anything and move on with fixing other stuff.
Again, this is a reason I want to get some testing in place before
making decisions like this; it is just hard to know if making a change
like this really will break anything (other than by doing smoke
testing, which is time consuming and itself error prone). We don't
need 100% coverage or anything; we just need a test harness so that we
can poke at specific parts of the code in a repeatable way before we
make changes to those same parts.

If we cannot change to errno.ENOSYS, then the better thing would be to
create better documentation for the behavior, the gotchas, and where
fusepy diverges from the default libfuse behavior. This keeps the same
existing userbase, while still being more inviting to new users.
Essentially making the code more transparent and discoverable.

I don't see an issue with deprecation. There are parts of the Python
standard library that have been deprecated for years, but never
removed (and still have no removal date). So long as there is another
option for users to migrate to, all it says is that there is something
better out there. We also don't necessarily need to raise deprecation
warnings in the code (although I don't have an issue with it, since
they don't do anything by default). Again, this is assuming that we go
the route of explicit deprecation at all.

I hope my half conscious babblings make some sense.

**Comment by mxmlnkn** [comment]
> As per the guidelines, it should either not exist or at least raise FuseOSError(errno.ENOSYS) (which does in fact work just fine).

Based on the [source code](https://github.com/libfuse/libfuse/blob/d04687923194d906fe5ad82dcd546c9807bf15b6/lib/fuse.c#L2104), this seems to work because the kernel, I assume(?), implements the fallback. Libfuse itself only tries to call the `create` callback and if it does not exist, then it returns `ENOSYS`.

However, this does not seem to be some universally consistent behavior. After adding the `read_buf`, which was added with libfuse 2.9, ratarmount fails when reading files because `read_buf` is used instead of `read` even when it returns `ENOSYS`. Looking at the [source code in libfuse](https://github.com/libfuse/libfuse/blob/d04687923194d906fe5ad82dcd546c9807bf15b6/lib/fuse.c#L1834-L1836) for the `read_buf` call, there is no fallback depending on the returned error. It simply calls `read_buf` if it is set or else `read`.

> Alternatively, one can subclass from Operations but, instead, has to overwrite (!) every single unwanted method and raise FuseOSError(errno.ENOSYS) when they are called.

The methods can also be set to `None` in the subclass. This is already done by fusepy for callbacks such as `bmap`.

> When offering a clean Operations class, it is probably also a good idea to "zero-out" the fuse_operations ctypes structure fuse_ops after its initialization and before it is filled with pointers to methods.

I think that this is unnecessary because the ctypes module should initialize everything. At least there are [tests](https://github.com/python/cpython/blob/f8a736b8e14ab839e1193cb1d3955b61c316d048/Lib/test/test_ctypes/test_numbers.py#L95) for this for primitive types in CPython.

> Because the base class code helps indicate to new users what methods
they need to implement to get a working FS. It is a form of passive
documentation. I know I looked at the `Operations` class
implementation in the code when I first tried to learn what to do with
the library.

I think that this is the only reason for those dummy implementations to exist. I have two ideas:
 - Simply move this "documentation" of the function signatures into the `Operations` class docstring.
 - Add [some kind of decorator that adds some kind of flag](https://tkjoetang.medium.com/python-decorator-use-it-as-feature-flag-b28f1c03bcc9) that makes it possible to detect dummy functions and ignore them. Methods that are `None` such as `bmap` are already ignored in such a manner with [this check](https://github.com/fusepy/fusepy/blob/5d997d6706cc0204e1b3ca679651485a7e7dda49/fuse.py#L681).

Another consideration: The libfuse debug log output, when enabled, is [only printed when callbacks for those functions are set](https://github.com/libfuse/libfuse/blob/d04687923194d906fe5ad82dcd546c9807bf15b6/lib/fuse.c#L2091). But, I guess, a user interested in such debug output should simply implement dummy returning ENOSYS for all those of interest.
